drivers/input/keyboard_upper.c

```diff
@@ -0,0 +1,435 @@
+  FAR struct keyboard_upperhalf_s *upper = inode->i_private;
+  FAR struct keyboard_lowerhalf_s *lower = upper->lower;
+
+  if (lower->write != NULL)
+    {
+      return lower->write(lower, buffer, buflen);
+    }
+
+  return -ENOSYS;
+}
+
+/****************************************************************************
+ * Public Function
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: keyboard_register
+ ****************************************************************************/
+
+int keyboard_register(FAR struct keyboard_lowerhalf_s *lower,
+                      FAR const char *path)
+{
+  FAR struct keyboard_upperhalf_s *upper;
+  int ret;
+
+  iinfo("Registering %s\n", path);
+  if (lower == NULL)
+    {
+      ierr("ERROR: invalid touchscreen device\n");
+      return -EINVAL;
+    }
+
+  upper = kmm_zalloc(sizeof(struct keyboard_upperhalf_s));
+  if (upper == NULL)
+    {
+      ierr("ERROR: Failed to mem alloc!\n");
+      return -ENOMEM;
+    }
+
+  upper->lower = lower;
+  lower->priv  = upper;
+  list_initialize(&upper->head);
+  nxsem_init(&upper->exclsem, 0, 1);
+
+  ret = register_driver(path, &g_keyboard_fops, 0666, upper);
+  if (ret < 0)
+    {
+      nxsem_destroy(&upper->exclsem);
+      kmm_free(upper);
+      return ret;
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: keyboard_unregister
+ ****************************************************************************/
+
+int keyboard_unregister(FAR struct keyboard_lowerhalf_s *lower,
+                        FAR const char *path)
+{
+  FAR struct keyboard_upperhalf_s *upper;
+  int ret;
+
+  DEBUGASSERT(lower != NULL);
+  DEBUGASSERT(lower->priv != NULL);
+
+  upper = lower->priv;
+
+  iinfo("UnRegistering %s\n", path);
+  ret = unregister_driver(path);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  nxsem_destroy(&upper->exclsem);
+  kmm_free(upper);
+  return 0;
+}
+
+/****************************************************************************
+ * keyboard_event
+ ****************************************************************************/
+
+void keyboard_event(FAR struct keyboard_lowerhalf_s *lower, uint32_t keycode,
+                    uint32_t type)
+{
+  FAR struct keyboard_upperhalf_s *upper = lower->priv;
+  FAR struct keyboard_opriv_s     *opriv;
+  struct keyboard_event_s          key;
+
+  if (nxsem_wait(&upper->exclsem) < 0)
+    {
+      return;
+    }
+
+  key.code = keycode;
+  key.type = type;
+  list_for_every_entry(&upper->head, opriv, struct keyboard_opriv_s, node)
+    {
+      if (nxsem_wait(&opriv->locksem) < 0)
+        {
+          continue;
+        }
+
+      circbuf_overwrite(&opriv->circ, &key, sizeof(struct keyboard_event_s));
+      nxsem_post(&opriv->waitsem);
+      keyboard_notify(opriv);
+      nxsem_post(&opriv->locksem);
```

Comment:
Minor optional: we can change this to
```suggestion
      if (nxsem_wait(&opriv->locksem) >= 0)
        {
          circbuf_overwrite(&opriv->circ, &key, sizeof(struct keyboard_event_s));
          nxsem_post(&opriv->waitsem);
          keyboard_notify(opriv);
          nxsem_post(&opriv->locksem);
        }
```